lib/utils/agentconn/agent_windows.go

```diff
@@ -20,20 +20,228 @@ limitations under the License.
 package agentconn
 
 import (
+	"encoding/binary"
+	"fmt"
 	"net"
+	"os"
+	"os/exec"
+	"regexp"
+	"strconv"
+	"strings"
 
 	"github.com/Microsoft/go-winio"
 	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+
+	apiutils "github.com/gravitational/teleport/api/utils"
 )
 
 const namedPipe = `\\.\pipe\openssh-ssh-agent`
 
 // Dial creates net.Conn to a SSH agent listening on a Windows named pipe.
 // This is behind a build flag because winio.DialPipe is only available on
-// Windows.
+// Windows. If connecting to a named pipe fails and we're in a Cygwin
+// environment, a connection to a Cygwin SSH agent will be attempted.
 func Dial(socket string) (net.Conn, error) {
 	conn, err := winio.DialPipe(namedPipe, nil)
 	if err != nil {
+		// MSYSTEM is used to specify what Cygwin environment is used;
+		// if it exists, there's a very good chance we're in a Cygwin
+		// environment
+		if msys, ok := os.LookupEnv("MSYSTEM"); ok && msys != "" {
+			conn, err := dialCygwin(socket)
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+			return conn, nil
+		}
+
+		return nil, trace.Wrap(err)
+	}
+
+	return conn, nil
+}
+
+const wellKnownSIDPrefix = "S-1-5-"
+
+var (
+	// format of the contents of a file created by Cygwin 'ssh-agent'
+	cygwinSocket = regexp.MustCompile(`!<socket >(\d+) (s )?([A-Fa-f0-9-]+)`)
+	// format of an output line from Cygwin 'ps'
```

Comment:
Can you include an example in the comment? These regexes are hard to grok by just looking at them.

Reply:
Done